Documentation for the new MP Config properties for MP Metrics 5.1 (#6959)
OpenLiberty/docs: modules/ROOT/pages/microprofile-config-properties.adoc

|`mp.metrics.appName`
|Passes the value of the  `_app` tag from the application server to the application. This property is used to avoid metric name collisions when multiple applications are deployed to the runtime. It is defined through a `microprofile-config.properties` file in the `META-INF` directory of the application. The value of this tag is appended to the output of each metric that originates from that application.

> In MicroProfile Metrics 5.0, the  `_app` tag that was defined in prior releases by the `mp.metrics.appName` MicroProfile Config property is designated as `mp_app`. When running in an MicroProfile Metrics 5.x environment, if the `mp.metrics.appName` property is defined, the `mp.metrics.defaultAppName` property must be defined as well.

Reviewer: ```suggestion
In MicroProfile Metrics 5.0 and later, the  `_app` tag that is defined in prior releases by the `mp.metrics.appName` MicroProfile Config property is designated as `mp_app`. In a MicroProfile Metrics 5.x environment, if the `mp.metrics.appName` property is defined, you must also define the `mp.metrics.defaultAppName` property. See the `mp.metrics.defaultAppName` property description for more information.
```

Reviewer: If you have to define both props, do they get the same value? If not, what is the relation between values?

Author: They will be different values. 
A vendor metric (provided by Open Liberty) can originate from the server side and application side.
For example, servlet metrics.

If the application created a servlet metric it can look like this:
`servlet_request_total{mp_scope="vendor",servlet="TestAppJakarta_sand_box_MyApplication",} 1.0`
The servlet created on the server side (i.e., hitting the `/metrics` endpoint can generate one will create one like this:
`servlet_request_total{mp_scope="vendor",servlet="io_openliberty_microprofile_metrics_5_0_public_internal_PublicMetricsRESTProxyServlet",} 4.0`

The tag set consists of `mp_scope` and `servlet`.

If the `mp.metrics.appName` is defined by an application (we'll give it a value of `myApp` it can create a metric like this:
`servlet_request_total{mp_app="myApp",mp_scope="vendor",servlet="TestAppJakarta_sand_box_MyApplication",} 1.0`

The servlet created on the server side (i.e., hitting the `/metrics` endpoint can generate one will create one like this:
`servlet_request_total{mp_scope="vendor",servlet="io_openliberty_microprofile_metrics_5_0_public_internal_PublicMetricsRESTProxyServlet",} 4.0`

At this point, the server side metric is missing the `mp_app`. This causes a mismatch in tags and will cause a registration error for the second metric to be registered.

The `mp.metrics.defaultAppName` serves to provide this `mp_app` tag for server side metrics (it can be any value they wish and does not have to match). The set of  tag names must match, but their values may differ.

Author: Perhaps:

In MicroProfile Metrics 5.0 and later, the  `_app` tag that is defined in prior releases by the `mp.metrics.appName` MicroProfile Config property is designated as `mp_app`. In a MicroProfile Metrics 5.x environment, if the `mp.metrics.appName` property is defined, you must also define the `mp.metrics.defaultAppName` property. See the `mp.metrics.defaultAppName` for more information.

Reviewer: Thanks- updated my suggestion.

|This value is read and evaluated during run time.
|`mp.metrics.appName=myApp`

|`mp.metrics.defaultAppName`

> |This MP config value is defined on the server level and is used in conjunction with the `mp.metrics.appName` config value when running under a MicroProfile Metrics 5.x environment. The value of `mp.metrics.defaultAppName` passes the value of the `mp_app` tag to metrics created on the server level. There exists vendor metrics (i.e., servlet metrics) that can be created both under the server and application context. Due to restriction introduced in MicroProfile Metrics 5.0 where the set of tag names must be the same, it is important to use the `mp.metrics.defaultAppName` with `mp.metrics.appName` name to avoid any registration exceptions.

Reviewer: ```suggestion
|The value for this property is defined at the server level and is used in conjunction with the `mp.metrics.appName` property when you run in a MicroProfile Metrics 5.x environment. The value of the  `mp.metrics.defaultAppName` property passes the value of the `mp_app` tag to metrics that are created on the server level. 

This property is necessary to use with the `mp.metrics.appName` property due to the restriction introduced in MicroProfile Metrics 5 .0 that metrics of the same name must contain the same set of tag names. Certain vendor metrics, such as servlet metrics, can create metrics in both the application and server context. The `mp.metrics.appName` property creates the `mp_app` tag for the servlet metrics that are created in the application context. However, without a similar tag value provided for servlet metrics that are created in the server context, a registration exception is thrown. The `mp.metrics.defaultAppName` property that is defined at the server level provides the necessary `mp_app` tag for metrics that are created at the server level. 
```

Author: See comment from https://github.com/OpenLiberty/docs/pull/6959/files#r1355764129

Author: Perhaps:

The value for this property is defined at the server level and is used in conjunction with the `mp.metrics.appName` property when you run in a MicroProfile Metrics 5.x environment. The value of the  `mp.metrics.defaultAppName` property passes the value of the `mp_app` tag to metrics that are created on the server level.

This property is necessary to use with the `mp.metrics.appName` property due to the restriction introduced in MicroProfile Metrics 5 .0 where metrics of the same name must contain the same set tag names. There exists certain vendor metrics, such as servlet metrics, that can create metrics in both the application and server context. The `mp.metrics.appName`  will create the `mp_app` tag for the servlet metrics created in the application context. However, without a similar tag value provided for servlet metrics created in the server context a registration exception will thrown. the `mp.metrics.defaultAppName` that is defined at the server level will provide the necessary `mp_app` value for metrics created at the server level.

Reviewer: updated my comment

|Server startup
|`mp.metrics.tags=app=shop,tier=integration,special=deli\=ver\,y`

|`mp.metrics.distribution.percentiles`

> a|Specifies a custom set of percentiles for matching Histogram and Timer metrics to track and output. Can be used to disable percentile output.

Reviewer: ```suggestion
a|Specifies a list of metric names that are paired with custom sets of percentiles for the metric to track and output. This property is applicable only to histogram and timer metrics.
```

> a|Specifies a custom set of percentiles for matching Histogram and Timer metrics to track and output. Can be used to disable percentile output.

Reviewer: > Specifies a custom set of percentiles for matching histogram and timer metrics to track and output

this is seems vague and Im not sure what to recommend... what is being tracked and what is being output?

Reviewer: and what are the "matching" metrics? Is it hist metrics that match timer metrics? Or hist and timer metrics that match across the server and app levels?

Author: The Histogram and Timer metric output data related to percentiles (of the recorded data).
There is a default set of percentiles (50th, 75th, 95th, 98th, 99th and 99.9th percentiles)
The property allows users to define their own custom set of percentiles that are tracked and then output.

Regarding the "Matching".
The properties follow a <property_name>=<metric_name1>=<values1>';<metric_name2>=<values2>';...;<metric_nameN>=<valuesN>'

The "matching" metric is matching the metric name defined in the property.
The bullet points were there to hopefully define what that meant (plus the example)

Reviewer: Ok- so the metrics that "match" with the metric name defined in the property. I think this could be easy to misread as saying the timer metrics match the histogram metrics. AT least that's what I first thought.

Would this then also be correct?

> Defines a custom set of percentiles for the specified histogram and timer metrics to track and output.

Author: Yep, that sounds more clear than what I had 😂

Author: Also, I guess we can probably remove the last sentence in favor of the last bullet point below:

`Setting the property without any value disables percentiles for all histogram and timer metrics.`

> * The property accepts a semi-colon separated set of values that consist of a metric name followed by an equals sign (`=`) and comma separated percentile values.

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that consist of a metric name followed by an equals sign (`=`) and comma-separated percentile values.
```

* The asterisk (`*`) can be used as a wildcard at the end of the metric name.

> * The percentile values are float values between 0.0 and 1.0 inclusively. Invalid values will be ignored.

Reviewer: ```suggestion
* The percentile values are float values between 0.0 and 1.0, inclusively. Invalid values are ignored.
```

* Setting the property with a metric name but no percentile values disables percentile output for that metric.
* Setting the property without any value disables percentiles for all histogram and timer metrics.

> * Values defined later take precedence over values before (i.e., right to left precedence).

Reviewer: ```suggestion
* Value precedence is right-to-left, so that values that are specified later in the property definition take precedence over previously-specified values.
```

This property is available starting in MicroProfile Metrics 5.1.
|Server startup
|`mp.metrics.distribution.percentiles=test.histogram=0.8,0.86,0.95;test.timer=0.9,0.99;test.*=0.5,0.6;other.*=`

|`mp.metrics.distribution.histogram.buckets`

> a| Specifies a custom set of histogram buckets for matching Histogram metrics to track and output.

Reviewer: see comment at line 292

Author: https://github.com/OpenLiberty/docs/pull/6959/files#r1355768990

Reviewer: added a suggestion based on our slack convo

> a| Specifies a custom set of histogram buckets for matching Histogram metrics to track and output.

Reviewer: ```suggestion
a| Specifies a list of metric names that are paired with custom sets of histogram buckets for the metric to track and output. This property is applicable only to histogram metrics.
```

> * The property accepts a semi-colon separated set of values that consist of a metric name followed by an equals sign (`=`) and comma separated bucket values:

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that consist of a metric name followed by an equals sign (`=`) and comma-separated bucket values.
```

> * Accepts a comma separated list of float and integer values greater than 0.

Reviewer: ```suggestion
* Accepts a comma-separated list of float and integer values greater than 0.
```

* The asterisk (`*`) can be used as a wildcard at the end of the metric name.

> * Defining a metric name with no bucket values or invalid values provides no effect.

Reviewer: ```suggestion
* Defining a metric name with no bucket values or invalid values has no effect.
```

> * Values defined later take precedence over values before (i.e., right to left precedence).

Reviewer: ```suggestion
Value precedence is right-to-left, so that values that are specified later in the property definition take precedence over previously-specified values.
```

This property is available starting in MicroProfile Metrics 5.1.
|Server startup
|`mp.metrics.distribution.histogram.buckets=test.alpha.histogram=10.0,50.0,100.0;test.beta.histogram=30.0,50.0,123`

|`mp.metrics.distribution.timer.buckets`

> a| Specifies a custom set of histogram buckets for matching Timer metrics to track and output.

Reviewer: see line 292

Author: https://github.com/OpenLiberty/docs/pull/6959/files#r1355768990

> * The property accepts a semi-colon separated set of values that consist of a metric name followed by an equals sign (`=`) and comma separated bucket values:

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that consist of a metric name followed by an equals sign (`=`) and comma-separated bucket values.
```

> * Accepts a comma separated list of integers with an appended time unit (valid time units are `ms` for milliseconds, `s` for seconds, `m` for minutes and `h` for hours). Float values will not be accepted. Values with no time unit default to milliseconds.

Reviewer: ```suggestion
* The property accepts a comma-separated list of integers with an appended time unit (valid time units are `ms` for milliseconds, `s` for seconds, `m` for minutes and `h` for hours). Float values will not be accepted. Values with no time unit default to milliseconds.
```

* The asterisk (`*`) can be used as a wildcard at the end of the metric name.

> * Defining a metric name with no bucket values or invalid values provides no effect.

Reviewer: ```suggestion
* Defining a metric name with no bucket values or invalid values has no effect.
```

> * Values defined later take precedence over values before (i.e., right to left precedence).

Reviewer: ```suggestion
* Value precedence is right-to-left, so that values that are specified later in the property definition take precedence over previously-specified values.
```

This property is available starting in MicroProfile Metrics 5.1.
|Server startup
|`mp.metrics.distribution.timer.buckets=test.*=50s,100s;test.alpha.timer=100,500ms,1s`

|`mp.metrics.distribution.percentiles-histogram.enabled`

> a| Specifies whether the the matching Histogram or Timer metric is to provide a default set of histogram buckets.

Reviewer: ```suggestion
a| Specifies whether the the matching histogram or timer metric provides a default set of histogram buckets.
```

Reviewer: what is being matched?

> * The property accepts a semi-colon separated set of values that consist of a metric name followed by an equals sign (`=`) and either `true` or `false.

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that consist of a metric name followed by an equals sign (`=`) and either `true` or `false.
```

* The asterisk (`*`) can be used as a wildcard at the end of the metric name.
* Defining a metric name with no values or invalid values has no effect.
* Values defined later take precedence over values before (i.e., right to left precedence).

This property is available starting in MicroProfile Metrics 5.1.
|Server startup
| `mp.metrics.distribution.percentiles-histogram.enabled=test.timer=true;test.histogram=false`

|`mp.metrics.distribution.histogram.min-value`

> a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured for a matching Histogram metric to provide a default set of histogram buckets, this MP config property is used to configure a lower bound for the set of buckets for the matching Histogram metric.

Reviewer: ```suggestion
a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a matching histogram metric , this property configures a lower bound for that default set of histogram buckets.
```

> * The property accepts a semi-colon separated set of values that define a metric name followed by an equals sign (`=`) and a value:

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that consist of a metric name followed by an equals sign (`=`) and a value.
```

* Properties accept float or integer values greater than 0.
* The asterisk (`*`) can be used as a wildcard at the end of the metric name.

> * Defining a metric name with no value or an invalid value provides no effect.

Reviewer: ```suggestion
* Defining a metric name with no value or an invalid value has no effect.
```

> * Values defined later take precedence over values before (i.e., right to left precedence).

Reviewer: ```suggestion
* Value precedence is right-to-left, so that values that are specified later in the property definition take precedence over previously-specified values.
```

This property is available starting in MicroProfile Metrics 5.1.
|Server startup
| `mp.metrics.distribution.histogram.min-value=test.histogram=50`

|`mp.metrics.distribution.histogram.max-value`

> a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured for a matching Histogram metric to provide a default set of histogram buckets, this MP config property is used to configure an upper bound for the set of buckets for the matching Histogram metric.

Reviewer: ```suggestion
a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a matching histogram metric , this property configures a upper bound for that default set of histogram buckets.
```

> * The property accepts a semi-colon separated set of values that define a metric name followed by an equals sign (`=`) and a value:

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that consist of a metric name followed by an equals sign (`=`) and a value:
```

* Properties accept float or integer values greater than 0.
* The asterisk (`*`) can be used as a wildcard at the end of the metric name.

> * Defining a metric name with no value or an invalid value provides no effect.

Reviewer: ```suggestion
* Defining a metric name with no value or an invalid value has no effect.
```

> * Values defined later take precedence over values before (i.e., right to left precedence).

Reviewer: ```suggestion
* Value precedence is right-to-left, so that values that are defined later in the property definition take precedence over earlier values.
```

This property is available starting in MicroProfile Metrics 5.1.
|Server startup
| `mp.metrics.distribution.histogram.max-value=test.histogram=120`

|`mp.metrics.distribution.timer.min-value`

> a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured for a matching Timer metric to provide a default set of histogram buckets, this MP config property is used to configure a lower bound for the set of buckets for the matching Timer metric.

Reviewer: ```suggestion
a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a matching Timer metric , this property configures a lower bound for that default set of histogram buckets.
```

> * The property accepts a semi-colon separated set of values that define a metric name followed by an equals sign (`=`) and a value:

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that consist of a metric name followed by an equals sign (`=`) and a value:
```

* Properties accept float or integer values greater than 0.
* The asterisk (`*`) can be used as a wildcard at the end of the metric name.

> * Defining a metric name with no value or an invalid value provides no effect.

Reviewer: ```suggestion
* Defining a metric name with no value or an invalid value has no effect.
```

> * Values defined later take precedence over values before (i.e., right to left precedence).

Reviewer: ```suggestion
* Value precedence is right-to-left, so that values that are specified later in the property definition take precedence over previously-specified values.
```

This property is available starting in MicroProfile Metrics 5.1.
|Server startup
| `mp.metrics.distribution.histogram.min-value=test.timer=200`

|`mp.metrics.distribution.timer.max-value`

> a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured for a matching Timer metric to provide a default set of histogram buckets, this MP config property is used to configure an upper bound for the set of buckets for the matching Timer metric.

Reviewer: ```suggestion
a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a matching Timer metric , this property configures an upper bound for that default set of histogram buckets.
```

> * The property accepts a semi-colon separated set of values that define a metric name followed by an equals sign (`=`) and a value:

Reviewer: ```suggestion
* The property accepts a semicolon-separated set of values that define a metric name followed by an equals sign (`=`) and a value:
```

* Properties accept float or integer values greater than 0.
* The asterisk (`*`) can be used as a wildcard at the end of the metric name.

> * Defining a metric name with no value or an invalid value provides no effect.

Reviewer: ```suggestion
* Defining a metric name with no value or an invalid value has no effect.
```

> * Values defined later take precedence over values before (i.e., right to left precedence).

Reviewer: ```suggestion
* Value precedence is right-to-left, so that values that are specified later in the property definition take precedence over previously-specified values.
```

|`mp.metrics.distribution.timer.buckets=test.*=50s,100s;test.alpha.timer=100,500ms,1s`

|`mp.metrics.distribution.percentiles-histogram.enabled`

> a| Specifies a list of metric names that are paired with a true or false value. This property is applicable only to histogram and timer metrics. If set to true, then the metric will output a default set of histogram buckets.

Reviewer: per IBM style- avoid future tense

```suggestion
a| Specifies a list of metric names that are paired with a true or false value. This property is applicable only to histogram and timer metrics. If set to true, the metric outputs a default set of histogram buckets.
```

| `mp.metrics.distribution.histogram.min-value=test.histogram=50`

|`mp.metrics.distribution.histogram.max-value`

> a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a histogram metric , this property configures a upper bound for that default set of histogram buckets.

Reviewer: ```suggestion
a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a histogram metric, this property configures a upper bound for that default set of histogram buckets.
```

| `mp.metrics.distribution.histogram.max-value=test.histogram=120`

|`mp.metrics.distribution.timer.min-value`

> a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a timer metric , this property configures a lower bound for that default set of histogram buckets.

Reviewer: ```suggestion
a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a timer metric, this property configures a lower bound for that default set of histogram buckets.
```

| `mp.metrics.distribution.histogram.min-value=test.timer=200`

|`mp.metrics.distribution.timer.max-value`

> a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a timer metric , this property configures an upper bound for that default set of histogram buckets.

Reviewer: ```suggestion
a| When the `mp.metrics.distribution.percentiles-histogram.enabled` is configured to provide a default set of histogram buckets for a timer metric, this property configures an upper bound for that default set of histogram buckets.
```
